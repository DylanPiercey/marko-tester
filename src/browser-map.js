const glob = require('glob');
const { existsSync } = require('fs');
const { relative, resolve } = require('path');
const {
  getModulePath,
  markoPath,
  markoWidgetsPath,
  markoVersion,
} = require('./versions');

function getPackages() {
  let marko3Packages = [];

  if (markoVersion === 3) {
    marko3Packages = [
      ...glob.sync(`${markoWidgetsPath}/!(node_modules)/**/package.json`),
      ...glob.sync(`${markoWidgetsPath}/package.json`),
      ...glob.sync(`${getModulePath('raptor-renderer')}/package.json`),
      ...glob.sync(`${getModulePath('raptor-dom')}/package.json`),
    ];
  }

  return [
    ...glob.sync(`${markoPath}/!(node_modules)/**/package.json`),
    ...glob.sync(`${markoPath}/!(node_modules)/package.json`),
    ...marko3Packages,
  ];
}

function getExisting(packagePath) {
  /* eslint-disable-next-line global-require, import/no-dynamic-require */
  const packageData = require(packagePath);
  const { main = '' } = packageData;
  let { browser = {} } = packageData;
  const relativeTo = `${packagePath.split('node_modules').slice(0, -1).join('node_modules')}/node_modules`;

  if (typeof browser === 'string') {
    browser = { [main]: browser };
  }

  return Object.entries(browser).reduce((cleanMap, [nodePath, browserPath]) => {
    /* eslint-disable-next-line no-param-reassign */
    browserPath = browserPath || resolve(__dirname, 'noop.js');

    const fullNodePath = resolve(packagePath, '..', nodePath);
    const fullBrowserPath = resolve(packagePath, '..', browserPath);

    return {
      ...cleanMap,
      ...(
        existsSync(fullNodePath) && existsSync(fullBrowserPath)
          ? { [relative(relativeTo, fullNodePath)]: relative(relativeTo, fullBrowserPath) }
          : {}
      ),
    };
  }, {});
}

function getBrowserMaps(packagePaths) {
  return packagePaths
    .reduce((map, packagePath) => ({
      ...map,
      ...getExisting(packagePath),
    }), {});
}

let browserMap;

try {
  browserMap = getBrowserMaps(getPackages());
} catch (error) {
  console.error(error);

  browserMap = {};
}

module.exports = browserMap;
