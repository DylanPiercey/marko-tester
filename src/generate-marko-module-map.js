const glob = require('glob');
const { isDebug } = require('marko/env');
const { promisify } = require('util');
const { readFile, writeFile, exists } = require('fs');
const { relative, resolve: resolvePath, join } = require('path');

const readFileAsync = promisify(readFile);
const writeFileAsync = promisify(writeFile);
const fileExistsAsync = promisify(exists);
const NODE_MODULES_PATH = relative(`${__dirname}/..`, resolvePath(require.resolve('marko'), '..', '..'));
const MARKO_DIST = `marko/${isDebug || process.env.NODE_ENV === 'test' ? 'src' : 'dist'}`;
const BASE_PATH = `${NODE_MODULES_PATH}/${MARKO_DIST}`;
const OUTPUT_PATH = './marko-modules-mocking-map.json';

function writeJSONFile(targetFilename, data) {
  return writeFileAsync(targetFilename, JSON.stringify(data, null, 2));
}

function notNull(item) {
  return !!item;
}

function getPackageFiles() {
  return new Promise((resolve, reject) => {
    glob(`${BASE_PATH}/**/package.json`, {}, (err, files) => {
      if (err) {
        reject(err);
      } else {
        resolve(files);
      }
    });
  });
}

function flattenBrowserMap(folderName, browserMap) {
  return Object.keys(browserMap).reduce((curr, mapKey) => {
    const mockSource = join(folderName, mapKey);
    const mockTarget = join(folderName, browserMap[mapKey]);

    return Object.assign(curr, {
      [mockSource]: mockTarget,
    });
  }, {});
}

async function processPackageFile(packageFile) {
  const text = await readFileAsync(packageFile, { encoding: 'utf8' });
  const contentJSON = JSON.parse(text);

  if (contentJSON.browser) {
    const key = packageFile.replace(BASE_PATH, MARKO_DIST).replace('/package.json', '');

    return flattenBrowserMap(key, contentJSON.browser);
  }

  return null;
}

async function filterExistingFiles(browserMap) {
  const existingModules = await Promise.all(Object.keys(browserMap).map(async (mapKey) => {
    const targetModule = join(NODE_MODULES_PATH, mapKey);
    const mockModule = join(NODE_MODULES_PATH, browserMap[mapKey]);
    const isTargeFileExisted = await fileExistsAsync(targetModule);
    const isMockFileExisted = await fileExistsAsync(mockModule);

    console.log(targetModule);
    console.log(mockModule);
    console.log(`Both files exist: ${isTargeFileExisted && isMockFileExisted ? '✅' : '❌'}`);

    return (isTargeFileExisted && isMockFileExisted) && ({
      [mapKey]: browserMap[mapKey],
    });
  }));

  return existingModules
    .filter(module => !!module)
    .reduce((curr, bMap) => Object.assign(curr, bMap), {});
}

(async function main() {
  try {
    const files = await getPackageFiles();
    const browserMaps = await Promise.all(files.map(processPackageFile));
    const mergedMap = browserMaps.filter(notNull)
      .reduce((curr, browserMap) => Object.assign(curr, browserMap), {});
    const resultMap = await filterExistingFiles(mergedMap);

    console.log('Generated module mocking map:');
    console.log(JSON.stringify(resultMap, null, 2));

    await writeJSONFile(OUTPUT_PATH, resultMap);
  } catch (err) {
    console.error('ERROR on generating map:', err);
  }
}());
