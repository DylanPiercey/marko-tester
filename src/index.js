const { readdirSync } = require('fs');
const {
  resolve, join, extname, basename,
} = require('path');
const { existsSync } = require('fs');
const stackTrace = require('stack-trace');
const browserMap = require('./browser-map');
const render = require('./render');

delete global.WeakMap;

Object.keys(browserMap)
  .forEach(moduleToMock => jest.mock(
    moduleToMock,
    () => require.requireActual(browserMap[moduleToMock]),
  ));

const getFullPath = (path) => {
  const stack = stackTrace.get();

  stack.splice(0, 2);

  const index = stack.findIndex((trace) => {
    const filename = trace.getFileName();
    const fullPath = resolve(filename || '', '..', path);

    return existsSync(fullPath) || existsSync(`${fullPath}.js`);
  });

  return index > -1 && resolve(stack[index].getFileName(), '..', path);
};
const getFixtures = (fixturesPath => () => (fixturesPath ? readdirSync(fixturesPath) : [])
  .filter(filename => /\.js(on)?$/.test(filename))
  .reduce((fixtures, filename) => {
    const absPath = join(fixturesPath, filename);
    const extension = extname(filename);
    const testName = basename(filename).replace(extension, '');

    /* eslint-disable-next-line global-require, import/no-dynamic-require */
    Object.assign(fixtures, { [testName]: require(absPath) || {} });

    return fixtures;
  }, {}))(getFullPath('__snapshots__'));
const runFixtures = (fixtures, fullPath, withAwait) => (fixtureName) => {
  const fixturesEntries = Object.entries(fixtures);
  const fixturesPath = getFullPath('__snapshots__');

  if (fixturesPath && fixturesEntries.length === 0) throw new Error(`No fixtures where found for component in "${fullPath}".`);

  fixturesEntries
    .filter(([name]) => (!fixtureName || fixtureName === name))
    .forEach(([name, fixture]) => it(`should render component with ${name} fixture`, async () => {
      const comp = await render(fullPath, withAwait)(JSON.parse(JSON.stringify(fixture)));

      expect(Array.from(document.body.childNodes)).toMatchSnapshot();
      comp.destroy();
    }));

  return {};
};
const helpers = {
  createEvent: (eventName, options = {}) => ((customEvent) => {
    const { bubbles = true, cancelable = true } = options;

    customEvent.initEvent(eventName, bubbles, cancelable);

    return Object.assign(customEvent, options);
  })(document.createEvent('Event')),
  defer: () => (toReturn => Object.assign(toReturn, {
    promise: new Promise((res, rej) => Object.assign(toReturn, { resolve: res, reject: rej })),
  }))({}),
};
const tester = (componentPath, { withoutFixtures, withAwait } = {}) => {
  const fullPath = getFullPath(componentPath);

  if (!fullPath) throw new Error(`Cannot find specified component at "${componentPath}".`);

  const fixtures = getFixtures();
  const boundRunFixtures = runFixtures(fixtures, fullPath, withAwait);

  return Object.assign(helpers, {
    fixtures: Object.assign(
      withoutFixtures ? boundRunFixtures : boundRunFixtures(),
      fixtures,
    ),
    render: render(fullPath, withAwait),
  });
};

module.exports = Object.assign(tester, helpers);
