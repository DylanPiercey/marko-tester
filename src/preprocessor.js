const { taglibFinder, taglibLookup, compileFile } = require('marko/compiler');
const { markoVersion } = require('./versions');

const shallowPatch = () => {
  if (markoVersion === 3) {
    try {
      taglibLookup.registerTaglib(require.resolve('marko-widgets/marko'));
    } catch (error) {
      console.warn('No marko-widgets were found.');
    }
  }

  taglibFinder.find = (a, b) => b;
};

module.exports = {
  process(src, filepath, { globals = {} }) {
    const { tester = {} } = globals;
    const { shallow = true } = tester;

    if (shallow) shallowPatch();

    return (markoVersion === 3)
      ? compileFile(filepath)
      /* eslint-disable-next-line global-require, import/no-unresolved */
      : require('marko/compiler').compileFileForBrowser(filepath, {
        ignoreUnrecognizedTags: true,
        escapeAtTags: true,
      });
  },
};
