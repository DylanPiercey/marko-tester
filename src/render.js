const clone = require('just-clone');
const { markoVersion } = require('./versions');

const mount = comp => comp.appendTo(document.body);
const getWidgetInstance = (renderer, input) => {
  if (/\.marko$/.test(renderer.path)) {
    document.body.insertAdjacentHTML('afterbegin', renderer.renderToString(clone(input)));

    return {
      destroy: () => document.body.childNodes.forEach(node => document.body.removeChild(node)),
    };
  }

  const mountWidget = comp => mount(comp).getWidget();

  return mountWidget(renderer.renderSync(clone(input)));
};
const getComponentInstance = (renderer, input, withAwait) => {
  const mountComponent = comp => mount(comp).getComponent();

  /* eslint-disable-next-line global-require, import/no-unresolved */
  require('marko/components').init();

  return withAwait
    ? renderer.render(clone(input)).then(mountComponent)
    : mountComponent(renderer.renderSync(clone(input)));
};

module.exports = (fullPath, withAwait) => (input) => {
  jest.resetModules();

  /* eslint-disable-next-line global-require, import/no-dynamic-require */
  const renderer = require(fullPath);

  return markoVersion === 3
    ? getWidgetInstance(renderer, input)
    : getComponentInstance(renderer, input, withAwait);
};
