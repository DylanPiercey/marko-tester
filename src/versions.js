const getModulePath = moduleName => require.resolve(`${moduleName}/package.json`).replace(/\/package\.json$/, '');
const markoPath = getModulePath('marko');
/* eslint-disable-next-line global-require, import/no-dynamic-require */
const { version: markoVersion } = require(`${markoPath}/package`);
let markoWidgetsPath;
let markoWidgetsVersion;

try {
  markoWidgetsPath = getModulePath('marko-widgets');
  /* eslint-disable-next-line global-require, import/no-dynamic-require */
  markoWidgetsVersion = require(`${markoWidgetsPath}/package`).version;
} catch (error) { /* */ }

module.exports = {
  getModulePath,
  markoPath,
  markoWidgetsPath,
  markoVersion: parseInt(markoVersion, 10),
  markoWidgetsVersion: markoWidgetsVersion && parseInt(markoWidgetsVersion, 10),
};
