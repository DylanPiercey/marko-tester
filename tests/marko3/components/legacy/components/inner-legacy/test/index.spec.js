require('../../../../../../../src/index')('../../../../../../marko4/components/legacy/components/inner-legacy/component.js');
