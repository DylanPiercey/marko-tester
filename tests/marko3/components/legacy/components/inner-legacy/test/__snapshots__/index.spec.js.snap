// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component with index fixture 1`] = `
Array [
  <div
    class="inner-legacy"
    id="wc0"
  >
    "world"
  </div>,
]
`;
